cpp-package/example/inference/imagenet_inference.cpp

```diff
@@ -0,0 +1,597 @@
+ * 4. use_gpu: determine if run inference on GPU
+ * 5. dataset: data file (.rec) to be used for inference
+ * 6. data_nthreads: number of threads for data loading
+ * 7. data_layer_type: data type for data layer
+ * 8. rgb_mean: mean value to be subtracted on R/G/B channel
+ * 9. rgb_std: standard deviation on R/G/B channel
+ * 10. shuffle_chunk_seed: shuffling chunk seed
+ * 11. seed: shuffling seed
+ * 12. benchmark: use dummy data for inference
+ *
+ * The constructor will:
+ *  1. Create ImageRecordIter based on the given dataset file.
+ *  2. Load the model and parameter files.
+ *  3. Infer and construct NDArrays according to the input argument and create an executor.
+ */
+Predictor::Predictor(const std::string& model_json_file,
+                     const std::string& model_params_file,
+                     const Shape& input_shape,
+                     bool use_gpu,
+                     const std::string& dataset,
+                     const int data_nthreads,
+                     const std::string& data_layer_type,
+                     const std::vector<float>& rgb_mean,
+                     const std::vector<float>& rgb_std,
+                     int shuffle_chunk_seed,
+                     int seed, bool benchmark)
+    : input_shape(input_shape),
+      use_gpu_(use_gpu),
+      dataset_(dataset),
+      data_nthreads_(data_nthreads),
+      data_layer_type_(data_layer_type),
+      rgb_mean_(rgb_mean),
+      rgb_std_(rgb_std),
+      shuffle_chunk_seed_(shuffle_chunk_seed),
+      seed_(seed),
+      benchmark_(benchmark) {
+  if (use_gpu) {
+    global_ctx = Context::gpu();
+  }
+
+  // initilize data iterator
+  if (!benchmark_ && !CreateImageRecordIter()) {
+    LG << "Error: failed to create ImageRecordIter";
+    throw std::runtime_error("ImageRecordIter cannot be created");
+  }
+
+  // Load the model
+  LoadModel(model_json_file);
+  // Initilize the parameters
+  // benchmark=false, load parameters from file
+  // benchmark=true, randomly initialize parameters
+  if (!benchmark_) {
+    LoadParameters(model_params_file);
+  } else {
+    InitParameters();
+  }
+
+  int dtype = GetDataLayerType();
+  if (dtype == -1) {
+    throw std::runtime_error("Unsupported data layer type...");
+  }
+  args_map["data"] = NDArray(input_shape, global_ctx, false, dtype);
+  Shape label_shape(input_shape[0]);
+  args_map["softmax_label"] = NDArray(label_shape, global_ctx, false);
+  std::vector<NDArray> arg_arrays;
+  std::vector<NDArray> grad_arrays;
+  std::vector<OpReqType> grad_reqs;
+  std::vector<NDArray> aux_arrays;
+
+  // infer and create ndarrays according to the given input ndarrays.
+  net.InferExecutorArrays(global_ctx, &arg_arrays, &grad_arrays, &grad_reqs, &aux_arrays, args_map,
+                      std::map<std::string, NDArray>(), std::map<std::string, OpReqType>(),
+                      aux_map);
+  for (auto& i : grad_reqs) i = OpReqType::kNullOp;
```

Comment:
In `InferExecutorArrays` function, if a user didn't specify `OpReqType` for an argument, it will be assigned a default value, like `kNullOp` for data and label, `kWriteTo` for others.  However, we don't need consider updating gradients in only forward pass, so we need reassign `kNullOp` for all arguments.